File: backend/static/onboarding.js
/**
 * HowYouBeen Onboarding Frontend
 * Handles the multi-step onboarding process
 */

// Global state
let currentStep = 'start';
let sessionId = null;
let availablePlatforms = [];
let connectedPlatforms = new Set();
let uploadedFiles = [];

// API base URL
const API_BASE = '/api/onboarding';

/**
 * Initialize the onboarding interface
 */
document.addEventListener('DOMContentLoaded', function() {
    loadAvailablePlatforms();
});

/**
 * Show error message
 */
function showError(message) {
    const errorDiv = document.getElementById('error-message');
    errorDiv.textContent = message;
    errorDiv.style.display = 'block';
    setTimeout(() => {
        errorDiv.style.display = 'none';
    }, 5000);
}

/**
 * Show success message in processing status
 */
function showProcessingStatus(message) {
    const statusDiv = document.getElementById('processing-status');
    statusDiv.innerHTML = `<p>${message}</p>`;
}

/**
 * Navigate to a specific step
 */
function goToStep(stepName) {
    // Hide current step
    const currentStepElement = document.querySelector('.step.active');
    if (currentStepElement) {
        currentStepElement.classList.remove('active');
    }
    
    // Show new step
    const newStepElement = document.getElementById(`step-${stepName}`);
    if (newStepElement) {
        newStepElement.classList.add('active');
        currentStep = stepName;
        
        // Update progress bar
        updateProgressBar(stepName);
    }
}

/**
 * Update progress bar
 */
function updateProgressBar(stepName) {
    const stepMap = {
        'start': 1,
        'basic-info': 2,
        'data-sources': 3,
        'visibility-config': 4,
        'processing': 5,
        'complete': 5
    };
    
    const stepNumber = stepMap[stepName] || 1;
    
    // Update progress steps
    for (let i = 1; i <= 5; i++) {
        const stepElement = document.getElementById(`step-${i}`);
        if (i < stepNumber) {
            stepElement.classList.add('completed');
            stepElement.classList.remove('active');
        } else if (i === stepNumber) {
            stepElement.classList.add('active');
            stepElement.classList.remove('completed');
        } else {
            stepElement.classList.remove('active', 'completed');
        }
    }
}

/**
 * Start the onboarding process
 */
async function startOnboarding() {
    try {
        const response = await fetch(`${API_BASE}/start`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            }
        });
        
        const data = await response.json();
        
        if (response.ok) {
            sessionId = data.session_id;
            goToStep('basic-info');
        } else {
            showError(data.detail || 'Failed to start onboarding');
        }
    } catch (error) {
        showError('Network error: ' + error.message);
    }
}

/**
 * Submit basic information
 */
async function submitBasicInfo() {
    const form = document.getElementById('basic-info-form');
    const formData = new FormData(form);
    
    const requestData = {
        session_id: sessionId,
        full_name: formData.get('full_name'),
        bio: formData.get('bio'),
        username: formData.get('username'),
        email: formData.get('email')
    };
    
    try {
        const response = await fetch(`${API_BASE}/basic-info`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(requestData)
        });
        
        const data = await response.json();
        
        if (response.ok) {
            goToStep('data-sources');
        } else {
            showError(data.detail || 'Failed to save basic information');
        }
    } catch (error) {
        showError('Network error: ' + error.message);
    }
}

/**
 * Load available platforms
 */
async function loadAvailablePlatforms() {
    try {
        const response = await fetch(`${API_BASE}/available-platforms`);
        const data = await response.json();
        
        if (response.ok) {
            availablePlatforms = data.platforms;
            renderPlatformsGrid(data.platforms, data.descriptions);
        }
    } catch (error) {
        console.error('Failed to load platforms:', error);
    }
}

/**
 * Render platforms grid
 */
function renderPlatformsGrid(platforms, descriptions) {
    const grid = document.getElementById('platforms-grid');
    grid.innerHTML = '';
    
    platforms.forEach(platform => {
        const card = document.createElement('div');
        card.className = 'platform-card';
        card.innerHTML = `
            <h4>${platform.charAt(0).toUpperCase() + platform.slice(1)}</h4>
            <p>${descriptions[platform] || 'Connect your account'}</p>
            <button onclick="connectPlatform('${platform}')" class="btn-secondary">
                ${connectedPlatforms.has(platform) ? 'Connected ✓' : 'Connect'}
            </button>
        `;
        
        if (connectedPlatforms.has(platform)) {
            card.classList.add('connected');
        }
        
        grid.appendChild(card);
    });
}

/**
 * Connect to a platform
 */
async function connectPlatform(platform) {
    if (connectedPlatforms.has(platform)) {
        return; // Already connected
    }
    
    try {
        const response = await fetch(`${API_BASE}/data-source`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                session_id: sessionId,
                platform: platform,
                credentials: {} // Mock credentials
            })
        });
        
        const data = await response.json();
        
        if (response.ok) {
            connectedPlatforms.add(platform);
            // Re-render grid to show connected state
            loadAvailablePlatforms();
            showProcessingStatus(`Successfully connected to ${platform}`);
        } else {
            showError(data.detail || `Failed to connect to ${platform}`);
        }
    } catch (error) {
        showError('Network error: ' + error.message);
    }
}

/**
 * Upload a file
 */
async function uploadFile() {
    const fileInput = document.getElementById('file-upload');
    const descriptionInput = document.getElementById('file-description');
    
    if (!fileInput.files.length) {
        showError('Please select a file to upload');
        return;
    }
    
    const file = fileInput.files[0];
    const description = descriptionInput.value;
    
    const formData = new FormData();
    formData.append('session_id', sessionId);
    formData.append('description', description);
    formData.append('file', file);
    
    try {
        const response = await fetch(`${API_BASE}/upload-document`, {
            method: 'POST',
            body: formData
        });
        
        const data = await response.json();
        
        if (response.ok) {
            uploadedFiles.push({
                filename: file.name,
                description: description,
                document_id: data.document_id
            });
            
            renderUploadedFiles();
            
            // Clear form
            fileInput.value = '';
            descriptionInput.value = '';
            
            showProcessingStatus(`Successfully uploaded ${file.name}`);
        } else {
            showError(data.detail || 'Failed to upload file');
        }
    } catch (error) {
        showError('Network error: ' + error.message);
    }
}

/**
 * Render uploaded files list
 */
function renderUploadedFiles() {
    const list = document.getElementById('uploaded-files-list');
    list.innerHTML = '';
    
    if (uploadedFiles.length === 0) {
        return;
    }
    
    const header = document.createElement('h4');
    header.textContent = 'Uploaded Files:';
    list.appendChild(header);
    
    uploadedFiles.forEach(file => {
        const fileDiv = document.createElement('div');
        fileDiv.className = 'uploaded-file';
        fileDiv.innerHTML = `
            <div>
                <strong>${file.filename}</strong>
                ${file.description ? `<br><small>${file.description}</small>` : ''}
            </div>
            <span style="color: #4CAF50;">✓</span>
        `;
        list.appendChild(fileDiv);
    });
}

/**
 * Submit visibility configuration
 */
async function submitVisibilityConfig() {
    const checkboxes = document.querySelectorAll('.visibility-categories input[type="checkbox"]:checked');
    const categories = [];
    
    checkboxes.forEach(checkbox => {
        categories.push({
            type: checkbox.value,
            name: checkbox.value === 'custom' ? 'Custom Category' : null
        });
    });
    
    if (categories.length === 0) {
        showError('Please select at least one visibility category');
        return;
    }
    
    try {
        const response = await fetch(`${API_BASE}/visibility-config`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                session_id: sessionId,
                categories: categories
            })
        });
        
        const data = await response.json();
        
        if (response.ok) {
            goToStep('processing');
        } else {
            showError(data.detail || 'Failed to configure visibility settings');
        }
    } catch (error) {
        showError('Network error: ' + error.message);
    }
}

/**
 * Process user data
 */
async function processUserData() {
    const processBtn = document.getElementById('process-btn');
    processBtn.disabled = true;
    processBtn.textContent = 'Processing...';
    
    showProcessingStatus('Starting AI profile generation...');
    
    try {
        // Simulate processing steps
        setTimeout(() => showProcessingStatus('Analyzing uploaded documents...'), 1000);
        setTimeout(() => showProcessingStatus('Connecting to data sources...'), 2000);
        setTimeout(() => showProcessingStatus('Generating your AI personality...'), 3000);
        setTimeout(() => showProcessingStatus('Finalizing your profile...'), 4000);
        
        const response = await fetch(`${API_BASE}/process`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                session_id: sessionId
            })
        });
        
        const data = await response.json();
        
        if (response.ok && data.success) {
            displayCompletionResults(data);
            goToStep('complete');
        } else {
            showError(data.error || 'Failed to process user data');
            processBtn.disabled = false;
            processBtn.textContent = 'Retry Processing';
        }
    } catch (error) {
        showError('Network error: ' + error.message);
        processBtn.disabled = false;
        processBtn.textContent = 'Retry Processing';
    }
}

/**
 * Display completion results
 */
function displayCompletionResults(data) {
    const completionDiv = document.getElementById('completion-details');
    
    completionDiv.innerHTML = `
        <div class="completion-card">
            <h3>Your AI Profile is Ready!</h3>
            <p>Profile URL: <strong>${data.profile_url}</strong></p>
            <p>User ID: <strong>${data.user_id}</strong></p>
        </div>
        
        <div class="ai-summary">
            <h4>AI Generated Summary:</h4>
            <p style="font-style: italic; padding: 20px; background: #f9f9f9; border-radius: 8px; margin: 20px 0;">
                "${data.ai_summary}"
            </p>
        </div>
        
        <div class="next-steps">
            <h4>Next Steps:</h4>
            <ul>
                ${data.next_steps.map(step => `<li>${step}</li>`).join('')}
            </ul>
        </div>
        
        <div style="margin-top: 30px;">
            <button onclick="viewProfile('${data.user_id}')" class="btn-primary">View Your Profile</button>
            <button onclick="startOver()" class="btn-secondary" style="margin-left: 15px;">Start Over</button>
        </div>
    `;
}

/**
 * View profile
 */
async function viewProfile(userId) {
    try {
        const response = await fetch(`${API_BASE}/user/${userId}`);
        const data = await response.json();
        
        if (response.ok) {
            alert(`Profile Data:\\n\\nUsername: ${data.username}\\nName: ${data.full_name}\\nBio: ${data.bio}\\nDiary Entries: ${data.diary_entries_count}\\nFacts: ${data.facts_count}\\nSources: ${data.sources_count}`);
        } else {
            showError('Failed to load profile data');
        }
    } catch (error) {
        showError('Network error: ' + error.message);
    }
}

/**
 * Start over
 */
function startOver() {
    if (confirm('Are you sure you want to start over? This will reset all your progress.')) {
        // Reset state
        sessionId = null;
        connectedPlatforms.clear();
        uploadedFiles = [];
        
        // Reset form
        document.getElementById('basic-info-form').reset();
        document.getElementById('file-description').value = '';
        document.getElementById('file-upload').value = '';
        document.getElementById('uploaded-files-list').innerHTML = '';
        
        // Uncheck visibility options
        document.querySelectorAll('.visibility-categories input[type="checkbox"]').forEach(cb => {
            cb.checked = ['close-family', 'best-friends', 'good-friends'].includes(cb.id);
        });
        
        // Go back to start
        goToStep('start');
        
        // Reload platforms
        loadAvailablePlatforms();
    }
}

File: backend/src/keepintouch/server/routes/onboarding.py
"""
Onboarding API routes

Handles the multi-step user onboarding process:
1. Start onboarding session
2. Submit basic info (name, bio)
3. Add data sources and upload documents  
4. Configure visibility categories
5. Process user data
6. Get status and results
"""

from fastapi import APIRouter, UploadFile, File, Form, HTTPException
from pydantic import BaseModel
from typing import Any, Dict, List, Optional

from ...ai_engine.onboarding_service import onboarding_service
from ...integrations.mock_services import mock_services
from ...storage.memory_store import store

router = APIRouter()


# Request/Response Models
class StartOnboardingResponse(BaseModel):
    session_id: str
    message: str


class BasicInfoRequest(BaseModel):
    session_id: str
    full_name: str
    bio: str
    username: str
    email: str


class BasicInfoResponse(BaseModel):
    success: bool
    message: str
    next_step: str


class DataSourceRequest(BaseModel):
    session_id: str
    platform: str
    credentials: Dict[str, str] = {}  # For demo, we don't need real credentials


class DataSourceResponse(BaseModel):
    success: bool
    message: str
    source_id: Optional[str] = None


class VisibilityConfigRequest(BaseModel):
    session_id: str
    categories: List[Dict[str, Any]]


class VisibilityConfigResponse(BaseModel):
    success: bool
    message: str


class ProcessRequest(BaseModel):
    session_id: str


class ProcessResponse(BaseModel):
    success: bool
    user_id: Optional[str] = None
    profile_url: Optional[str] = None
    ai_summary: Optional[str] = None
    next_steps: List[str] = []
    error: Optional[str] = None


@router.post("/start", response_model=StartOnboardingResponse)
async def start_onboarding():
    """Start a new onboarding session"""
    try:
        session_id = await onboarding_service.start_onboarding()
        return StartOnboardingResponse(
            session_id=session_id,
            message="Onboarding session started successfully"
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to start onboarding: {str(e)}")


@router.post("/basic-info", response_model=BasicInfoResponse)
async def submit_basic_info(request: BasicInfoRequest):
    """Submit basic user information (step a & b)"""
    try:
        success = await onboarding_service.submit_basic_info(
            session_id=request.session_id,
            name=request.full_name,
            bio=request.bio,
            username=request.username,
            email=request.email
        )
        
        if not success:
            raise HTTPException(status_code=400, detail="Failed to save basic info. Username may already exist.")
        
        return BasicInfoResponse(
            success=True,
            message="Basic information saved successfully",
            next_step="data_sources"
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error saving basic info: {str(e)}")


@router.post("/data-source", response_model=DataSourceResponse)
async def add_data_source(request: DataSourceRequest):
    """Add a data source connection (part of step c)"""
    try:
        info_source = await onboarding_service.add_data_source(
            session_id=request.session_id,
            platform=request.platform,
            credentials=request.credentials
        )
        
        if not info_source:
            raise HTTPException(status_code=400, detail=f"Failed to connect to {request.platform}")
        
        return DataSourceResponse(
            success=True,
            message=f"Successfully connected to {request.platform}",
            source_id=info_source.source_id
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error connecting data source: {str(e)}")


@router.post("/upload-document")
async def upload_document(
    session_id: str = Form(...),
    description: str = Form(""),
    file: UploadFile = File(...)
):
    """Upload a document (part of step c)"""
    try:
        # Read file content
        file_content = await file.read()
        
        document = await onboarding_service.upload_document(
            session_id=session_id,
            file_content=file_content,
            filename=file.filename or "unknown",
            description=description
        )
        
        if not document:
            raise HTTPException(status_code=400, detail="Failed to upload document")
        
        return {
            "success": True,
            "message": f"Document '{file.filename}' uploaded successfully",
            "document_id": document.document_id
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error uploading document: {str(e)}")


@router.post("/visibility-config", response_model=VisibilityConfigResponse)
async def configure_visibility(request: VisibilityConfigRequest):
    """Configure visibility categories (step d)"""
    try:
        success = await onboarding_service.configure_visibility(
            session_id=request.session_id,
            visibility_categories=request.categories
        )
        
        if not success:
            raise HTTPException(status_code=400, detail="Failed to configure visibility settings")
        
        return VisibilityConfigResponse(
            success=True,
            message="Visibility categories configured successfully"
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error configuring visibility: {str(e)}")


@router.post("/process", response_model=ProcessResponse)
async def process_user_data(request: ProcessRequest):
    """Process all user data and generate profile (step e)"""
    try:
        result = await onboarding_service.process_user_data(request.session_id)
        
        if not result.get("success"):
            return ProcessResponse(
                success=False,
                error=result.get("error", "Unknown error during processing")
            )
        
        return ProcessResponse(
            success=True,
            user_id=result.get("user_id"),
            profile_url=result.get("profile_url"),
            ai_summary=result.get("ai_summary"),
            next_steps=result.get("next_steps", [])
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error processing user data: {str(e)}")


@router.get("/status/{session_id}")
async def get_onboarding_status(session_id: str):
    """Get current onboarding status"""
    try:
        status = await onboarding_service.get_onboarding_status(session_id)
        
        if not status:
            raise HTTPException(status_code=404, detail="Onboarding session not found")
        
        return status
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error getting status: {str(e)}")


@router.get("/available-platforms")
async def get_available_platforms():
    """Get list of available data source platforms"""
    try:
        platforms = mock_services.get_available_platforms()
        return {
            "platforms": platforms,
            "descriptions": {
                "goodreads": "Connect your reading list and book reviews",
                "google_photos": "Import recent photos and memories",
                "linkedin": "Add professional profile information",
                "github": "Include your code repositories and projects",
                "instagram": "Share recent social media updates"
            }
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error getting platforms: {str(e)}")


@router.get("/user/{user_id}")
async def get_user_profile(user_id: str):
    """Get user profile (for testing)"""
    try:
        user = await store.get_user(user_id)
        
        if not user:
            raise HTTPException(status_code=404, detail="User not found")
        
        return {
            "user_id": user.user_id,
            "username": user.username,
            "full_name": user.full_name,
            "bio": user.bio,
            "onboarding_completed": user.onboarding_completed,
            "diary_entries_count": len(user.diary_entries),
            "facts_count": len(user.facts),
            "sources_count": len(user.sources)
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error getting user profile: {str(e)}")


@router.delete("/session/{session_id}")
async def cleanup_session(session_id: str):
    """Clean up onboarding session (for testing)"""
    try:
        session = await store.get_onboarding_session(session_id)
        if session:
            # In a real implementation, this would properly delete the session
            del store.onboarding_sessions[session_id]
            return {"message": "Session cleaned up successfully"}
        else:
            raise HTTPException(status_code=404, detail="Session not found")
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error cleaning up session: {str(e)}")